### DISCUSSION TITLE
Do the AoTuV and Lancer Patches Improve the Vorbis Encoder?

### DISCUSSION THREAD

**Discussion by ShamanZeno92** [enzo1982/freac#436] [Audio Format Discussion]
My understanding is that Fre:ac makes use of the aoTuV beta 6.03 and associated Lancer patch to improve sound quality and encoding speed. However, I cannot find any public listening tests to verify that aoTuV beta 6.03 actually improves anything, could just as easily introduce regressions. Documentation for the associated Lancer patch is also sparse. How much does it speed up encoding on modern hardware, and at what cost? Now, I don't claim to have golden ears, but Vorbis files encoded with the basic oggenc sound the same as those encoded by Fre:ac to me and my ears. Additionally files are a few KiB smaller when made by oggenc. As for speed, Fre:ac encoding was faster, but only by a few seconds on the tests files I encoded. Then again, I'm using a 10 year old PC. Essentially, I'm wondering if the inclusion of aoTuV and Lancer patches in Fre:ac was made under the assumption they improve the unpatched Vorbis encoder rather than being based on any documented proof or testing they do in fact make real improvements.

**Comment by enzo1982** [comment]
aoTuV has been developed between 2004 and 2011. Lots of discussions and evaluations of its benefits happened back then.

Many of those resources may no longer be online, but you might find some old discussions over at [HydrogenAudio](https://hydrogenaud.io/). I won't dig up the old threads for you, though.

The consensus back then was that aoTuV is superior to upstream Vorbis. So much so that some changes from aoTuV have been integrated into the official libvorbis. Later aoTuV changes became more complex, though, and may not have been integrated into the official library in order to keep the reference code simple.

Files encoded with aoTuV are usually a bit larger than those encoded with the official libvorbis. This is because aoTuV is supposed to allocate additional bits to areas where upstream libvorbis struggles to maintain the quality target.

If you want to challenge the consensus that aoTuV is superior to upstream Vorbis, you should conduct your own listening tests to back that up.

The Lancer patches implement SSE optimizations to improve encoding performance. They should not affect quality.

Compared to aoTuV, the Lancer patches improve performance by approximately 25% on my i7-13700KF system. The aoTuV version does more work than reference libvorbis, though, so it itself is about 15% slower than the reference code. That's why aoTuV+Lancer is only slightly faster than upstream libvorbis.

**Reply by ShamanZeno92** [reply]
That makes sense. However, my understanding was the quality numbers used in Vorbis encoders all represent the same perceived audio quality. So regardless of what encoder is used the quality for any given target q-number will remain the same, only that improved encoders will create files of the same quality with fewer bits. As mention [here](https://xiph.org/vorbis/faq/#_quality). Which would make AoTuV encoded files inferior because they are larger, using more bits to create the same audio quality. Unless of course, the AoTuV encoder's quality numbers represent different levels of audio quality than the reference encoder, or that, as you said, the upstream libvorbis struggles to meet its target quality.

**Reply by Lee-Carre** [reply]
> in order to keep the reference code simple.

That (it's not much of a reference if it's not easy to read & grok), and possibly other reasons I can imagine (yes, granted, pure speculation; the answer might be somewhere in a Xiph mailing list archive).

A reference implementation, to my understanding, is supposed to be
* stable (thus experimental changes are questionable)
* demonstrate correctness; in this case of implementing the encoding algorithms, bitstream structure, and such
* demonstrate robustness; not try to be clever, and certainly not prematurely optimised (especially since encoding is done much less than decoding, and hardware is continuing to improve; so optimal performance isn't crucial, it just needs to be reasonable on typical hardware; correctness of the data output matters much more) — if I recall, FLAC versioning isn't based on revisions to the execution code, but the format of the data which it outputs (since that's the important thing, not (within reason) how it's generated).
* ideally, be deterministic; for repeatable testing, and comparison
* less robustness (simplicity & transparency) tends to invite bugs to take up residence; exactly what one doesn't want in an implementation treated as the ideal (in terms of accuracy/correctness, if not performance)
* simplicity, in order to better preserve cleanness, modularity, and future extensibility

So, yeah, for a marginal gain, it doesn't seem worth it. Other implementations can experiment to find more efficient ways of encoding audio, while still outputting a playable bitstream. The reference implementation isn't the place for doing that. Only after experimenting & testing has been done, and yielded an obvious enhancement.

Developers/maintainers may have been keen to focus on newer codecs, especially once they considered Vorbis refined enough to be usable. Why flog a dying horse, when a fresh one awaits?

---

I'm reminded of something similar in the SETI@home project (of Berkeley University, California's Space Science Lab). Once the project had migrated to the BOINC middle-ware platform (for a whole bunch of reasons), some volunteers/enthusiasts produced a tweaked build of the source-code of the data-crunching part (which analyses work units of radio data, within the BOINC infrastructure).

As far as I'm aware, the changes were never merged upstream into the main build which was distributed by default. To use the modified build, required manual configuration-file editing of the BOINC client).

Changes in the modified build included:
* commenting out all the code related to the screen-saver
* commenting out various optional code, such as frequent status-reporting, and the like
* compiling with SSE(2) instruction set extensions enabled (there were a few different compilations, I recall, and you had to figure out the right one for your CPU (by checking CPU-Z, or similar)
* essentially, trimming it down in order that it would process a work unit as fast as possible


I think the SSE changes might have eventually  made it into the main codebase, along with code to auto-detect the CPU and then auto-select the optimal code branch (using different instruction set extensions); so that it was the same binary for everyone (per platform), but figured out the details for itself, and yielding closer-to-optimal performance.

From the PoV of managing a science project (as opposed to a tech/software project), the scientists & technicians were interested in getting faster results (but never at the cost of accuracy/correctness of processing; fast but sloppy analysis was useless to them) for least effort (the proverbial low-hanging fruit, as it were). I guess using enhanced instruction sets was deemed worth the trouble, whereas the other changes were not.

There was also, I think, a requirement for deterministic behaviour; the same work unit was sent out to a quorum of volunteer computers, in order to compare results to have reasonable assurance of correct processing, and thus accurate analysis answers (basically, were there any interesting signal candidates in that chunk of radio data (centred around the hydrogen alpha line, I think), or not (most returned a negative result, as you might imagine).

And now that Arecibo has collapsed/self-destructed, there's no work units for anyone for a while. Heh.
